frame/crowdloan-rewards/src/tests.rs

```diff
@@ -388,3 +401,91 @@ fn test_valid_eth_hardcoded() {
 		assert_eq!(CrowdloanRewards::claimed_rewards(), CrowdloanRewards::total_rewards());
 	});
 }
+
+#[test]
+fn valid_associate_transactions_are_free() {
+	with_rewards_default(|_, accounts| {
+		assert_ok!(CrowdloanRewards::initialize(Origin::root()));
+
+		for (reward_account, remote_account) in accounts {
+			let prevalidate_association = PrevalidateAssociation::<Test>::new();
+
+			let proof = remote_account.proof(reward_account.clone());
+
+			let call = Call::CrowdloanRewards(crate::Call::associate {
+				reward_account: reward_account.clone(),
+				proof,
+			});
+			let dispatch_info = call.get_dispatch_info();
+
+			assert_eq!(dispatch_info.pays_fee, Pays::No);
+			let validate_result =
+				prevalidate_association.validate(&reward_account, &call, &dispatch_info, 0);
+			assert_eq!(validate_result, TransactionValidity::Ok(ValidTransaction::default()));
+		}
+	});
+}
+
+#[test]
+fn already_associated_associate_transactions_are_recognized() {
+	with_rewards_default(|_, accounts| {
+		assert_ok!(CrowdloanRewards::initialize(Origin::root()));
+
+		for (reward_account, remote_account) in accounts.clone() {
+			assert_ok!(CrowdloanRewards::associate(
+				Origin::root(),
+				reward_account.clone(),
+				remote_account.proof(reward_account.clone()),
+			));
+		}
+
+		for (reward_account, remote_account) in accounts {
+			let prevalidate_association = PrevalidateAssociation::<Test>::new();
+
+			let proof = remote_account.proof(reward_account.clone());
+
+			let call = Call::CrowdloanRewards(crate::Call::associate {
+				reward_account: reward_account.clone(),
+				proof,
+			});
+			let dispatch_info = call.get_dispatch_info();
+
+			let validate_result =
+				prevalidate_association.validate(&reward_account, &call, &dispatch_info, 0);
+			assert_eq!(
+				validate_result,
+				TransactionValidity::Err(
+					InvalidTransaction::Custom(ValidityError::AlreadyAssociated as u8).into()
+				)
+			);
+		}
+	});
+}
+
+#[test]
+fn no_reward_associate_transactions_are_recognized() {
```

Comment:
quick question - if there is test for signed exstrinsic?

Reply:
Not sure what you mean - these are the tests for the SignedExtension

Reply:
Improved tests in latest commit, if that's what you meant